### PR TITLE
EDM-3840: Add sync infrastructure for dependency auto-sync

### PR DESCRIPTION
**Pull request by galel12**
## EDM-3840: Add sync infrastructure for dependency auto-sync

**Jira:** [EDM-3840](https://redhat.atlassian.net/browse/EDM-3840)

### Summary

Adds the foundational infrastructure for dependency auto-sync (EDM-3841):

- **`sync_state` table** — tracks external dependency fingerprints per org, keyed by `(org_id, resource_key)`
- **`dependency_refs` table** — maps fleets/devices to their external dependencies (git, HTTP, secret)
- **`dependenciesSync` API field** — optional on `FleetSpec` and `DeviceSpec`, with `enabled` and `pollInterval` (validated: 30s–24h)
- **`GitLsRemote`** — resolves git refs to commit SHAs without cloning, with credential sanitization

### Testing

- Unit tests for `DependenciesSync` validation (boundary values, defaults, Fleet/DeviceSpec integration)
- Unit tests for `GitLsRemote` error paths and credential redaction
- Integration tests for `sync_state` and `dependency_refs` store operations

### CODE REVIEW

File: internal/tasks/git_helpers.go [hunk 1]
```diff
@@ -87,6 +90,82 @@ func CloneGitRepo(repo *domain.Repository, revision *string, depth *int, cfg *co
 	return mfs, hash, nil
 }
 
+// GitLsRemote resolves a git ref (branch name, tag name, or full ref path)
+// to its commit SHA without cloning the repository. Error messages are
+// sanitized to prevent credential leakage.
+func GitLsRemote(ctx context.Context, repoURL string, ref string, auth transport.AuthMethod) (string, error) {
```
Comment (asafbennatan) [inline internal/tasks/git_helpers.go]: should accept a list of refs not a single one

File: internal/store/dependencyref.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	"github.com/flightctl/flightctl/internal/store/model"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+type DependencyRef interface {
+	InitialMigration(ctx context.Context) error
+	Upsert(ctx context.Context, orgID uuid.UUID, ref *model.DependencyRef) error
+	ListByFleet(ctx context.Context, orgID uuid.UUID, fleetName string) ([]model.DependencyRef, error)
```
Comment (asafbennatan) [inline internal/store/dependencyref.go]: i dont think these list methods will be used , also the delets , i think a delete function will be used (delete by key)

File: internal/store/syncstate.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+func (s *SyncStateStore) getDB(ctx context.Context) *gorm.DB {
+	return s.dbHandler.WithContext(ctx)
+}
+
+func (s *SyncStateStore) InitialMigration(ctx context.Context) error {
+	return s.getDB(ctx).AutoMigrate(&model.SyncState{})
+}
+
+func (s *SyncStateStore) Get(ctx context.Context, orgID uuid.UUID, resourceKey string) (*model.SyncState, error) {
```
Comment (asafbennatan) [inline internal/store/syncstate.go]: these get and set method wont be used ,
in general we should only commit methods that we will be using

File: internal/tasks/git_helpers.go (hunk 1 above)
Comment (galel12) [reply internal/tasks/git_helpers.go]: Fixed, updated in a0ecc3ec

File: internal/store/dependencyref.go (hunk 2 above)
Comment (galel12) [reply internal/store/dependencyref.go]: Fixed, trimmed to Upsert + Delete by key only. Updated in a0ecc3ec

File: internal/store/syncstate.go (hunk 3 above)
Comment (galel12) [reply internal/store/syncstate.go]: Fixed, trimmed to InitialMigration only. Updated in a0ecc3ec

File: test/integration/store/syncstate_test.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	)
+
+	BeforeEach(func() {
+		ctx = testutil.StartSpecTracerForGinkgo(suiteCtx)
+		log = flightlog.InitLogs()
+		storeInst, cfg, dbName, _ = store.PrepareDBForUnitTests(ctx, log)
+
+		orgId := uuid.New()
+		err := testutil.CreateTestOrganization(ctx, storeInst, orgId)
+		Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (galel12) [reply test/integration/store/syncstate_test.go:32]: @hferber1  - heads up on a guideline vs. codebase mismatch in [test/AGENTS.md](https://github.com/flightctl/flightctl/blob/main/test/AGENTS.md?plain=1#L35).

[Line 35](https://github.com/flightctl/flightctl/blob/main/test/AGENTS.md?plain=1#L35) states: "Unit/integration: Prefer table-driven tests; use testify/require"

However, every integration test in test/integration/store/ (e.g. [device_test.go](https://github.com/flightctl/flightctl/blob/main/test/integration/store/device_test.go), [fleet_test.go](https://github.com/flightctl/flightctl/blob/main/test/integration/store/fleet_test.go), [repository_test.go](https://github.com/flightctl/flightctl/blob/main/test/integration/store/repository_test.go)) uses Ginkgo/Gomega — the entire store suite runs through a Ginkgo TestStore runner. This came up during a CodeRabbit review on PR #2913 where it flagged our new store tests for using Gomega instead of testify/require.

Should [test/AGENTS.md](https://github.com/flightctl/flightctl/blob/main/test/AGENTS.md?plain=1#L35) be updated to acknowledge that store integration tests use Ginkgo/Gomega, or is there an intent to migrate them to testify/require? Happy to update the guideline if it's just a documentation gap.

File: api/core/v1beta1/validation_test.go [hunk 5]
```diff
@@ -2320,3 +2320,182 @@ func TestRepository_Validate_BackwardCompatibility(t *testing.T) {
...
+			if tt.wantErr {
+				found := false
+				for _, err := range errs {
+					if strings.Contains(err.Error(), tt.errSubstring) {
+						found = true
+						break
+					}
+				}
+				require.True(found, "expected %q error in: %v", tt.errSubstring, errs)
+			}
```
[the remark this answers is not in the document]
Comment (galel12) [reply api/core/v1beta1/validation_test.go:2458]: Fixed, added `require.Empty(errs)` on success paths. Updated in 8199f463

File: test/integration/store/syncstate_test.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	})
```
[the remark this answers is not in the document]
Comment (galel12) [reply test/integration/store/syncstate_test.go:33]: Not applicable — `PrepareDBForUnitTests` returns `(Store, *config.Config, string, *gorm.DB)`, not an error. The discarded 4th value is a `*gorm.DB` handle. Setup failures are handled internally via `log.Fatalf`.

File: test/integration/store/syncstate_test.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+	AfterEach(func() {
+		store.DeleteTestDB(ctx, log, cfg, storeInst, dbName)
+	})
+
+	Context("When running initial migration", func() {
+		It("should create the sync_state table without error", func() {
+			Expect(storeInst.SyncState()).ToNot(BeNil())
+		})
```
[the remark this answers is not in the document]
Comment (galel12) [reply test/integration/store/syncstate_test.go:42]: CRUD methods were intentionally removed per reviewer feedback (only InitialMigration kept for this story). Migration is verified by RunMigrations succeeding during test DB setup. CRUD will be added and tested when the sync controller story lands.

File: internal/store/dependencyref.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	"context"
+	"fmt"
+
+	"github.com/flightctl/flightctl/internal/store/model"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+type DependencyRef interface {
```
Comment (asafbennatan) [inline internal/store/dependencyref.go:13]: lets avoid creating the store functions until they are used , keep inital migration only

File: internal/tasks/git_helpers_test.go [hunk 9]
```diff
@@ -80,3 +84,104 @@ var _ = Describe("ConvertFileSystemToIgnition", func() {
 		})
 	})
 })
+
+func TestSanitizeGitError(t *testing.T) {
```
Comment (asafbennatan) [inline internal/tasks/git_helpers_test.go:88]: happy path ( a real working repo with refs) is not tested,
suggestion - ls remote on flightctl for example , if you consider this to be an integration test add that as an integration test

File: internal/store/dependencyref.go (hunk 8 above)
Comment (galel12) [reply internal/store/dependencyref.go:13]: Fixed, trimmed to InitialMigration only. Updated in 622eee37

File: internal/tasks/git_helpers_test.go (hunk 9 above)
Comment (galel12) [reply internal/tasks/git_helpers_test.go:88]: Added happy-path tests against flightctl public repo — single ref and multi-ref resolution. Guarded with testing.Short() skip. Updated in 622eee37

**Review by asafbennatan — APPROVED** [review]
LGTM

**Review by asafbennatan — APPROVED** [review]
LGTM
